### ISSUE TITLE
Add port parameter to DNS Secondary Server setup

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5468]
|<img alt="BobNovas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| BobNovas reported|
|-|-|
|Trac ID|trac#5468|
|Type|enhancement|
|Created|2012-01-31 00:27:37Z|
|Branch|https://github.com/twisted/twisted/tree/secondary-dns-server-alternate-port-5468|

Patch allows a DNS Secondary server to use a primary DNS server that is serving on a port other than port 53.

Attachments:

* [02-add Port parameter to DNS Secondary Server.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e8d/e8dd6463fc60247ef056a211346d4631430ed2ef/274d00bfe7f8ca5afbae2d7eb690b7fec54e05f5.patch) (2618 bytes) - added by BobNovas on 2012-01-31 00:27:55Z - 
* [alternate-port-5468-v01.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e8d/e8dd6463fc60247ef056a211346d4631430ed2ef/32c6d60cd9593016fd6055011c4e1e90d84ce2db.patch) (2099 bytes) - added by BobNovas on 2012-03-06 00:23:47Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5468 5468
type__enhancement enhancement
reporter__BobNovas BobNovas
priority__normal normal
milestone__ 
branch__branches_secondary_dns_server_alternate_port_5468 branches/secondary-dns-server-alternate-port-5468
branch_author__Bob_Novas__exarkun Bob Novas, exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1327969657000000 1327969657000000
changetime__1331043308000000 1331043308000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.

  1. Some general things:
    1. Please put import statements at the top of modules.
    1. Test method docstrings should be in the style "X does y." not "Test that X does y."
    1. All methods should be separated by two blank lines.  Classes should be separated by three blank lines.
  1. More interestingly:
    1. `_cancelDelayedCalls` defeats the purpose of the checking that the test runner is doing.  You should never write a unit test that cancels all timed events (or disconnects all file descriptors, or anything else like that).  Clean up the resources allocated in the test.  If there are more things left after that, it's a bug in the implementation to fix.
    1. The new values supported by the `primary` argument/attribute of `SecondaryAuthority` needs to be documented.  Also, as a matter of style, I would probably do the extra parsing just once in `__init__` rather than on each call to `transfer`.  And, perhaps more significantly than that, "parsing" is really not what you want to do here.  Just accept a tuple.  (I'd really like to suggest endpoints here, but we have no datagram endpoints yet.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [33620]) Branching to 'secondary-dns-server-alternate-port-5468'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [33621]) Apply 02-add\ Port\ parameter\ to\ DNS\ Secondary\ Server.patch 

refs [#5468](https://github.com/twisted/twisted/issues/5468)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Sorry, I missed a very important review comment:

In the tests, this code:

```
        self.assertEqual(d.callbacks[0][0][1][0].host, '192.168.1.1')
        self.assertEqual(d.callbacks[0][0][1][0].port, 53)
```

is not how you should ever unit test... anything.

    1. The particular callbacks on a `Deferred` should be considered an implementation detail
    1. The shape of the `callbacks` list on a `Deferred` is somewhat opaque; you shouldn't look at arguments in it.
    1. These assertions don't make a compelling argument that the code is behaving as desired, only that some attributes... ''somewhere''... got set.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [33622]) Move the ip:port parsing to the options layer, where it belongs.  Add lots of tests.

refs [#5468](https://github.com/twisted/twisted/issues/5468)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [33623]) also, these

refs [#5468](https://github.com/twisted/twisted/issues/5468)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

I pushed the parsing out to the tap module.  I also added alternate constructors to the two secondary-related classes, leaving the existing APIs alone.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/secondary-dns-server-alternate-port-5468)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set owner to BobNovas|
|-|-|

Thanks for working on this, JP, and for contributing all this functionality, Bob.

   1. test_constructAuthorityFromHostAndPort has an incorrect docstring (still talks about default port).
   1. Since I don't know much about DNS, a comment explaining why in test_transfer you don't decode first two bytes would be helpful (length-encoding?).
   1. The changes to `twisted/names/tap.py` have no tests.
   1. Perhaps the reactor should be explicitly passable to `SecondaryAuthority` via the public API? We are trying to move towards world where reactors are explicit rather than implicit.
   1. Docstring of `SecondaryAuthorityTests` is inaccurate description of what it's testing (it seems to refer to code in tap).

Please address these comments (via a patch against the branch) and put back in review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

> The changes to twisted/names/tap.py have no tests. 

It does, in a new file (so maybe you didn't notice it, due to the excellent behavior of `svn diff` to omit such things).

**Comment by twisted-trac** [comment]
|<img alt="BobNovas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|BobNovas removed owner|
|-|-|

This change addresses points 1,2 and 5. For point 3, there are tests for tap.py in test_tap.py, as exarkun noted above. Point 4 seems to me to be out of scope of this change.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst set owner to @itamarst|
|-|-|

Oops, sorry I missed test_tap.py. Looks good now, I will run tests on buildbot and merge if all goes well.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set status to `closed`|
|-|-|

(In [33655]) Merge secondary-dns-server-alternate-port-5468: Non-standard port support in secondary authorities.

Author: BobNovas, exarkun
Review: exarkun, itamar
Fixes: [#5468](https://github.com/twisted/twisted/issues/5468)

Secondary DNS authorities can now retrieve information from primaries with non-standard ports.
